Bug: Assertion raises INTERNALERROR with options --tb=line --junitxml=junit.xml

ReprEntry throws INTERNALERROR if you attempt to assert (with message) within a test.
Bug is encountered on pytest version 6.0 and above with options **--tb=line --junitxml=junit.xml**
(Not encountered on versions 5.4.3 and below)

### To Reproduce:
Run
`pytest --tb=line --junitxml=junit.xml`
for the following test:
```python
def test_ReprEntry_Assertion():
    assert False, 'Triggers INTERNALERROR from TerminalWriter'
```

**Observed output:**
```
============================================================== test session starts ===============================================================
platform linux -- Python 3.6.8, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /c/Users/parijatm/CodeBase/pytest
plugins: cov-2.8.1, timeout-1.3.4
collected 1 item                                                                                                                                 

test_tiny.py F
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/python3.6/site-packages/_pytest/main.py", line 240, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "/python3.6/site-packages/_pytest/main.py", line 296, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/python3.6/site-packages/pluggy/hooks.py", line 286, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/python3.6/site-packages/pluggy/manager.py", line 93, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/python3.6/site-packages/pluggy/manager.py", line 87, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "/python3.6/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/python3.6/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/python3.6/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/python3.6/site-packages/_pytest/main.py", line 321, in pytest_runtestloop
INTERNALERROR>     item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
INTERNALERROR>   File "/python3.6/site-packages/pluggy/hooks.py", line 286, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/python3.6/site-packages/pluggy/manager.py", line 93, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/python3.6/site-packages/pluggy/manager.py", line 87, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "/python3.6/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/python3.6/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/python3.6/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/python3.6/site-packages/_pytest/runner.py", line 100, in pytest_runtest_protocol
INTERNALERROR>     runtestprotocol(item, nextitem=nextitem)
INTERNALERROR>   File "/python3.6/site-packages/_pytest/runner.py", line 117, in runtestprotocol
INTERNALERROR>     reports.append(call_and_report(item, "call", log))
INTERNALERROR>   File "/python3.6/site-packages/_pytest/runner.py", line 211, in call_and_report
INTERNALERROR>     hook.pytest_runtest_logreport(report=report)
INTERNALERROR>   File "/python3.6/site-packages/pluggy/hooks.py", line 286, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/python3.6/site-packages/pluggy/manager.py", line 93, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/python3.6/site-packages/pluggy/manager.py", line 87, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "/python3.6/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/python3.6/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/python3.6/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/python3.6/site-packages/_pytest/junitxml.py", line 606, in pytest_runtest_logreport
INTERNALERROR>     reporter.append_failure(report)
INTERNALERROR>   File "/python3.6/site-packages/_pytest/junitxml.py", line 227, in append_failure
INTERNALERROR>     fail.append(bin_xml_escape(report.longrepr))
INTERNALERROR>   File "/python3.6/site-packages/_pytest/junitxml.py", line 82, in bin_xml_escape
INTERNALERROR>     return py.xml.raw(illegal_xml_re.sub(repl, py.xml.escape(str(arg))))
INTERNALERROR>   File "/python3.6/site-packages/_pytest/_code/code.py", line 905, in __str__
INTERNALERROR>     self.toterminal(tw)
INTERNALERROR>   File "/python3.6/site-packages/_pytest/_code/code.py", line 951, in toterminal
INTERNALERROR>     element[0].toterminal(tw)
INTERNALERROR>   File "/python3.6/site-packages/_pytest/_code/code.py", line 981, in toterminal
INTERNALERROR>     entry.toterminal(tw)
INTERNALERROR>   File "/python3.6/site-packages/_pytest/_code/code.py", line 1079, in toterminal
INTERNALERROR>     self._write_entry_lines(tw)
INTERNALERROR>   File "/python3.6/site-packages/_pytest/_code/code.py", line 1050, in _write_entry_lines
INTERNALERROR>     + "\n".join(self.lines)
INTERNALERROR> AssertionError: Unexpected failure lines between source lines:
INTERNALERROR> E   AssertionError: Triggers INTERNALERROR from TerminalWriter
INTERNALERROR>     assert False

========================================================== 1 failed, 1 warning in 0.38s ==========================================================
```


Issue appears to be with the 'not' in the following assertion:
https://github.com/pytest-dev/pytest/blob/91dbdb60930e1b573de366cbe7be6dac77ba200f/src/_pytest/_code/code.py#L1059

--- a/changelog/7707.bugfix.rst
+++ b/changelog/7707.bugfix.rst
@@ -1,0 +1,1 @@
+Fix internal error when handling some exceptions that contain multiple lines or the style uses multiple lines (``--tb=line`` for example).

--- a/src/_pytest/_code/code.py
+++ b/src/_pytest/_code/code.py
@@ -1049,25 +1049,21 @@ class ReprEntry(TerminalRepr):
         # such as ">   assert 0"
         fail_marker = "{}   ".format(FormattedExcinfo.fail_marker)
         indent_size = len(fail_marker)
-        indents = []
-        source_lines = []
-        failure_lines = []
-        seeing_failures = False
-        for line in self.lines:
-            is_source_line = not line.startswith(fail_marker)
-            if is_source_line:
-                assert not seeing_failures, (
-                    "Unexpected failure lines between source lines:\n"
-                    + "\n".join(self.lines)
-                )
+        indents = []  # type: List[str]
+        source_lines = []  # type: List[str]
+        failure_lines = []  # type: List[str]
+        for index, line in enumerate(self.lines):
+            is_failure_line = line.startswith(fail_marker)
+            if is_failure_line:
+                # from this point on all lines are considered part of the failure
+                failure_lines.extend(self.lines[index:])
+                break
+            else:
                 if self.style == "value":
                     source_lines.append(line)
                 else:
                     indents.append(line[:indent_size])
                     source_lines.append(line[indent_size:])
-            else:
-                seeing_failures = True
-                failure_lines.append(line)
 
         tw._write_source(source_lines, indents)
 

--- a/testing/code/test_excinfo.py
+++ b/testing/code/test_excinfo.py
@@ -4,6 +4,8 @@ import os
 import queue
 import sys
 import textwrap
+from typing import Any
+from typing import Dict
 from typing import Tuple
 from typing import Union
 
@@ -1045,28 +1047,34 @@ raise ValueError()
     @pytest.mark.parametrize(
         "reproptions",
         [
-            {
-                "style": style,
-                "showlocals": showlocals,
-                "funcargs": funcargs,
-                "tbfilter": tbfilter,
-            }
-            for style in ("long", "short", "no")
+            pytest.param(
+                {
+                    "style": style,
+                    "showlocals": showlocals,
+                    "funcargs": funcargs,
+                    "tbfilter": tbfilter,
+                },
+                id="style={},showlocals={},funcargs={},tbfilter={}".format(
+                    style, showlocals, funcargs, tbfilter
+                ),
+            )
+            for style in ["long", "short", "line", "no", "native", "value", "auto"]
             for showlocals in (True, False)
             for tbfilter in (True, False)
             for funcargs in (True, False)
         ],
     )
-    def test_format_excinfo(self, importasmod, reproptions):
-        mod = importasmod(
-            """
-            def g(x):
-                raise ValueError(x)
-            def f():
-                g(3)
-        """
-        )
-        excinfo = pytest.raises(ValueError, mod.f)
+    def test_format_excinfo(self, reproptions: Dict[str, Any]) -> None:
+        def bar():
+            assert False, "some error"
+
+        def foo():
+            bar()
+
+        # using inline functions as opposed to importasmod so we get source code lines
+        # in the tracebacks (otherwise getinspect doesn't find the source code).
+        with pytest.raises(AssertionError) as excinfo:
+            foo()
         file = io.StringIO()
         tw = TerminalWriter(file=file)
         repr = excinfo.getrepr(**reproptions)

